src/main/java/com/commercetools/sync/commons/helpers/GraphQlRequest.java

```diff
@@ -0,0 +1,129 @@
+package com.commercetools.sync.commons.helpers;
+
+import com.commercetools.sync.commons.models.GraphQlQueryEndpoint;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.sphere.sdk.client.HttpRequestIntent;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.client.SphereRequest;
+import io.sphere.sdk.http.HttpMethod;
+import io.sphere.sdk.http.HttpResponse;
+import io.sphere.sdk.json.SphereJsonUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+
+public class GraphQlRequest implements SphereRequest<GraphQlResult> {
+    protected final Set<String> keysToSearch;
+    protected final GraphQlQueryEndpoint endpoint;
+    private long limit = 500;
+    private String queryPredicate = null;
+
+    /**
+     * A SphereRequest implementation to allow {@link SphereClient} to execute graphQL queries on CTP. It provides a
+     * POST request to the CTP graphql API containing body to fetch a set of ids matching given keys of a resource
+     * defined in endpoint parameter.
+     *
+     * @param keysToSearch - a set of keys to fetch matching ids for.
+     * @param endpoint - a string representing the name of the resource endpoint.
+     */
+    public GraphQlRequest(@Nonnull final Set<String> keysToSearch, @Nonnull final GraphQlQueryEndpoint endpoint) {
+
+        this.keysToSearch = requireNonNull(keysToSearch);
+        this.endpoint = endpoint;
+    }
+
+    /**
+     * This method adds a predicate string to the request.
+     *
+     * @param predicate - a string representing a query predicate.
+     * @return - an instance of this class.
+     */
+    @Nonnull
+    public GraphQlRequest withPredicate(final String predicate) {
+
+        this.queryPredicate = predicate;
+        return this;
+    }
+
+    /**
+     * This method adds a limit to the request.
+     *
+     * @param limit - a number representing the query limit parameter
+     * @return - an instance of this class
+     */
+    @Nonnull
+    public GraphQlRequest withLimit(final long limit) {
+
+        this.limit = limit;
+        return this;
+    }
+
+    @Nullable
+    @Override
+    public GraphQlResult deserialize(final HttpResponse httpResponse) {
+
+        final JsonNode rootJsonNode = SphereJsonUtils.parse(httpResponse.getResponseBody());
+        if (rootJsonNode.isNull()) {
+            return null;
+        }
+        JsonNode result = rootJsonNode.get("data").get(endpoint.getName());
```

Comment:
I also think it should not increase. But we had some bad experience with this before and I'm not sure how this complexity thing is calculated and what aspects are involved, thus the note :) Maybe you can ask support if the complexity could ever be over 20k and what are the criteria for the complexity and what is the answer from the server when the complexity is too high?

Reply:
@salander85 can you ask the support how the graphql complexity is calculated and if it could be different from project to project with the same query?

Reply:
@salander85 From the answer of support, it seems like it could differ from a project to project. But I'm not sure if they understood that we use the same query for different projects? In their additional answer they said it depends on the field.

```
can complexity different from project to project?

Generally, yes, complexity can differ.  It depends on how you build your queries, which can be analyzed:

Our schema defines a "cost" per field, which you can take advantage of by analyzing your queries, using one of the two possibilities:

Inspecting the 
x-graphql-query-complexity response header and its value.

Use the GraphiQL's 
Profiling functionality to measure the impact.
```

Reply:
Anyway, I guess we can merge :) LGTM.